custom_components/unifiprotect/sensor.py

```diff
@@ -157,25 +297,62 @@ def __init__(
         self,
         protect: ProtectApiClient,
         protect_data: UnifiProtectData,
-        device: ProtectAdoptableDeviceModel,
+        device: ProtectDeviceModel,
         description: UnifiProtectSensorEntityDescription,
     ):
         """Initialize an Unifi Protect sensor."""
         self.entity_description: UnifiProtectSensorEntityDescription = description
         super().__init__(protect, protect_data, device, description)
         self._attr_name = f"{self.device.name} {self.entity_description.name}"
 
-    @property
-    def native_value(self) -> StateType | date | datetime:
-        """Return the state of the sensor."""
-        value = get_nested_attr(self.device, self.entity_description.ufp_value)
-
-        assert isinstance(value, (str, int, float, datetime)) or value is None
-        if isinstance(value, datetime) and not above_ha_version(2021, 12):
-            return value.replace(microsecond=0).isoformat()
+    @callback
+    def _async_nvr_value(self) -> float:
+        assert isinstance(self.device, NVR)
+        if self.entity_description.key == _KEY_MEMORY:
+            memory = self.device.system_info.memory
+            value = (1 - memory.available / memory.total) * 100
+        else:
+            record_type, subtype = self.entity_description.key.split("_")
+            if record_type == "record":
+                record_type_attr = "recording_type_distributions"
+                subtype_attr = "recording_type"
+            else:
+                record_type_attr = "resolution_distributions"
+                subtype_attr = "resolution"
 
+            distributions = self.device.storage_stats.storage_distribution
+            for item in getattr(distributions, record_type_attr):
+                if getattr(item, subtype_attr) == subtype:
+                    value = getattr(item, "percentage")
+                    break
         return value
 
+    @callback
+    def _async_update_device_from_protect(self) -> None:
+        super()._async_update_device_from_protect()
+
+        if self.entity_description.ufp_value is None:
+            new_value = self._async_nvr_value()
+        else:
+            new_value = get_nested_attr(self.device, self.entity_description.ufp_value)
+
+        if isinstance(new_value, timedelta):
+            new_value = new_value.total_seconds()
+
+        if isinstance(new_value, datetime):
+            # protection to prevent uptime from changing if UniFi Protect changes values slightly
+            if hasattr(self, "_attr_native_value") and self._attr_native_value:
+                diff = abs((self._attr_native_value - new_value).total_seconds())
+                if diff < 5:
+                    return
```

Comment:
Can we prevent the wobbly time from getting into the integration by cleaning this in up in `pyunifiprotect`?

Reply:
Not in a great way. The way keys are currently being filtered only has the incoming value, not the existing. By the time you can compare existing to new, the `changed_data` has already been generated for the subscription message which would cause a diff between `changed_data` and `new_obj`.

Reply:
I could also just truncate the seconds, but then that will lose a bit of resolution into when it came online.

Reply:
That seems like a good compromise for now.

Reply:
Just to clarify, to truncate the seconds?

Reply:
Yes